extensions/ql-vscode/src/common/vscode/file-path-discovery.ts

```diff
@@ -0,0 +1,190 @@
+  Uri,
+  WorkspaceFoldersChangeEvent,
+  workspace,
+} from "vscode";
+import { MultiFileSystemWatcher } from "./multi-file-system-watcher";
+import { AppEventEmitter } from "../events";
+import { extLogger } from "..";
+import { FilePathSet } from "../file-path-set";
+import { exists, lstat } from "fs-extra";
+import { containsPath } from "../../pure/files";
+import { getOnDiskWorkspaceFoldersObjects } from "./workspace-folders";
+
+interface PathData {
+  path: string;
+}
+
+/**
+ * Discovers all files matching a given filter contained in the workspace.
+ *
+ * Scans the whole workspace on startup, and then watches for changes to files
+ * to do the minimum work to keep up with changes.
+ *
+ * Can configure which changes it watches for, which files are considered
+ * relevant, and what extra data to compute for each file.
+ */
+export abstract class FilePathDiscovery<T extends PathData> extends Discovery {
+  /** The set of known paths we are tracking */
+  protected paths: T[] = [];
+  protected readonly onDidChangePathsEmitter: AppEventEmitter<void>;
+
+  private readonly changedFilePaths = new FilePathSet();
+  private readonly watcher: MultiFileSystemWatcher = this.push(
+    new MultiFileSystemWatcher(),
+  );
+
+  /**
+   * @param name Name of the discovery operation, for logging purposes.
+   * @param fileWatchPattern Passed to `vscode.RelativePattern` to determine the files to watch for changes to.
+   */
+  constructor(name: string, private readonly fileWatchPattern: string) {
+    super(name, extLogger);
+
+    this.onDidChangePathsEmitter = this.push(new EventEmitter<void>());
+    this.push(
+      workspace.onDidChangeWorkspaceFolders(
+        this.workspaceFoldersChanged.bind(this),
+      ),
+    );
+    this.push(this.watcher.onDidChange(this.fileChanged.bind(this)));
+  }
+
+  /**
+   * Compute any extra data to be stored regarding the given path.
+   */
+  protected abstract getDataForPath(path: string): Promise<T>;
+
+  /**
+   * Is the given path relevant to this discovery operation?
+   */
+  protected abstract pathIsRelevant(path: string): boolean;
+
+  /**
+   * Should the given new data overwrite the existing data we have stored?
+   */
+  protected abstract shouldOverwriteExistingData(
+    newData: T,
+    existingData: T,
+  ): boolean;
+
+  /**
+   * Do the initial scan of the entire workspace and set up watchers for future changes.
+   */
+  public async initialRefresh() {
+    getOnDiskWorkspaceFoldersObjects().forEach((workspaceFolder) => {
+      this.changedFilePaths.addPath(workspaceFolder.uri.fsPath);
+    });
+
+    this.updateWatchers();
+    return this.refresh();
+  }
+
+  private workspaceFoldersChanged(event: WorkspaceFoldersChangeEvent) {
+    event.added.forEach((workspaceFolder) => {
+      this.changedFilePaths.addPath(workspaceFolder.uri.fsPath);
+    });
+    event.removed.forEach((workspaceFolder) => {
+      this.changedFilePaths.addPath(workspaceFolder.uri.fsPath);
+    });
```

Comment:
Why are we adding paths for workspace folders which were removed? Shouldn't we be removing those instead?

Reply:
The idea here is just to register that "something has changed at this path" so then in the refresh we'll look at this path and update our internal state to match whatever we find on disk. I'll add a comment as to why we're doing the same thing for both added and removed paths.

I don't think it's worth it to actually update `this.paths` in this method. Then we'd have to worry about calling the event emitter. Adding to `this.changedFilePaths` and calling `refresh` does all that for us.